nomos-blend/message/src/encap.rs

```diff
@@ -0,0 +1,481 @@
+use std::iter::repeat_n;
+
+use nomos_core::wire;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    crypto::{
+        pseudo_random_bytes, Ed25519PrivateKey, Ed25519PublicKey, ProofOfQuota, ProofOfSelection,
+        SharedKey, Signature, X25519PrivateKey, KEY_SIZE, PROOF_OF_QUOTA_SIZE,
+        PROOF_OF_SELECTION_SIZE, SIGNATURE_SIZE,
+    },
+    error::Error,
+    message::{BlendingHeader, Header, Payload, PayloadHeader, PayloadType, PublicHeader},
+};
+
+// TODO: Consider having these as parameters or generics.
+const MAX_ENCAPSULATIONS: usize = 3;
+const MAX_PAYLOAD_BODY_SIZE: usize = 34 * 1024;
+
+/// An encapsulated message that is sent to the blend network.
+//
+// TODO: Implement [`BlendMessage`] trait for this type.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct EncapsulatedMessage {
+    /// A header that is not encapsulated.
+    header: Header,
+    /// A public header that is not encapsulated.
+    public_header: PublicHeader,
+    /// An encapsulated private header.
+    private_header: [Vec<u8>; MAX_ENCAPSULATIONS],
+    /// A encapsulated payload.
+    payload: Vec<u8>,
+}
+
+impl EncapsulatedMessage {
+    /// Encapsulates the given payload with the provided encapsulation inputs.
+    /// Encapsulation inputs more than [`MAX_ENCAPSULATIONS`] are not allowed.
+    pub fn new(
+        inputs: &[EncapsulationInput],
+        payload_type: PayloadType,
+        payload_body: &[u8],
+    ) -> Result<Self, Error> {
+        if inputs.len() > MAX_ENCAPSULATIONS {
```

Comment:
Do you think we could encode all these constraints into its own type, e.g., `RangeVec<const MIN: usize, const MAX: usize>`? Or something like that? Might be out of scope for this PR but wanted to have your opinion on this.